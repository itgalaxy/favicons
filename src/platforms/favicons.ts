import { FaviconHtmlElement } from "../index";
import { FaviconOptions, NamedIconOptions } from "../config/defaults";
import { transparentIcon, transparentIcons } from "../config/icons";
import { OptionalMixin, Platform, uniformIconOptions } from "./base";

const ICONS_OPTIONS: (NamedIconOptions & OptionalMixin)[] = [
  { name: "favicon.ico", ...transparentIcons(16, 24, 32, 48, 64) },
  { name: "favicon-16x16.png", ...transparentIcon(16) },
  { name: "favicon-32x32.png", ...transparentIcon(32) },
  { name: "favicon-48x48.png", ...transparentIcon(48) },
  { name: "favicon.svg", ...transparentIcon(1024), optional: true }, // arbitrary size. if more than one svg source is given, the closest to this size will be picked.
];

export class FaviconsPlatform extends Platform {
  constructor(options: FaviconOptions) {
    super(
      options,
      uniformIconOptions(options, options.icons.favicons, ICONS_OPTIONS),
    );
  }

  override async createHtml(): Promise<FaviconHtmlElement[]> {
    return this.iconOptions.map(({ name, ...options }) => {
      if (name.endsWith(".ico")) {
        // prettier-ignore
        return `<link rel="icon" type="image/x-icon" href="${this.cacheBusting(this.relative(name))}">`;
      } else if (name.endsWith(".svg")) {
        // prettier-ignore
        return `<link rel="icon" type="image/svg+xml" href="${this.cacheBusting(this.relative(name))}">`;
      }

      const { width, height } = options.sizes[0];

      // prettier-ignore
      return `<link rel="icon" type="image/png" sizes="${width}x${height}" href="${this.cacheBusting(this.relative(name))}">`;
    });
  }
}
