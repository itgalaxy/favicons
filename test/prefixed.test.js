import favicons from "../src";
import { logo_png } from "./util";

test("should allow setting an URL prefix", async () => {
  expect.assertions(1);

  const result = await favicons(logo_png, {
    path: "https://domain/subdomain",
  });

  await expect(result).toMatchFaviconsSnapshot();
});
