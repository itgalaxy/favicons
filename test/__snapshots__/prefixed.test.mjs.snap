// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should allow setting an URL prefix 63`] = `
{
  "files": [
    {
      "contents": "{
  "name": null,
  "short_name": null,
  "description": null,
  "dir": "auto",
  "lang": "en-US",
  "display": "standalone",
  "orientation": "any",
  "start_url": "/?homescreen=1",
  "background_color": "#fff",
  "theme_color": "#fff",
  "icons": [
    {
      "src": "https://domain/subdomain/android-chrome-36x36.png",
      "sizes": "36x36",
      "type": "image/png",
      "purpose": "any"
    },
    {
      "src": "https://domain/subdomain/android-chrome-48x48.png",
      "sizes": "48x48",
      "type": "image/png",
      "purpose": "any"
    },
    {
      "src": "https://domain/subdomain/android-chrome-72x72.png",
      "sizes": "72x72",
      "type": "image/png",
      "purpose": "any"
    },
    {
      "src": "https://domain/subdomain/android-chrome-96x96.png",
      "sizes": "96x96",
      "type": "image/png",
      "purpose": "any"
    },
    {
      "src": "https://domain/subdomain/android-chrome-144x144.png",
      "sizes": "144x144",
      "type": "image/png",
      "purpose": "any"
    },
    {
      "src": "https://domain/subdomain/android-chrome-192x192.png",
      "sizes": "192x192",
      "type": "image/png",
      "purpose": "any"
    },
    {
      "src": "https://domain/subdomain/android-chrome-256x256.png",
      "sizes": "256x256",
      "type": "image/png",
      "purpose": "any"
    },
    {
      "src": "https://domain/subdomain/android-chrome-384x384.png",
      "sizes": "384x384",
      "type": "image/png",
      "purpose": "any"
    },
    {
      "src": "https://domain/subdomain/android-chrome-512x512.png",
      "sizes": "512x512",
      "type": "image/png",
      "purpose": "any"
    }
  ]
}",
      "name": "manifest.webmanifest",
    },
    {
      "contents": "<?xml version="1.0" encoding="utf-8"?>
<browserconfig>
  <msapplication>
    <tile>
      <square70x70logo src="https://domain/subdomain/mstile-70x70.png"/>
      <square150x150logo src="https://domain/subdomain/mstile-150x150.png"/>
      <wide310x150logo src="https://domain/subdomain/mstile-310x150.png"/>
      <square310x310logo src="https://domain/subdomain/mstile-310x310.png"/>
      <TileColor>#fff</TileColor>
    </tile>
  </msapplication>
</browserconfig>",
      "name": "browserconfig.xml",
    },
    {
      "contents": "{
  "version": "1.0",
  "api_version": 1,
  "layout": {
    "logo": "https://domain/subdomain/yandex-browser-50x50.png",
    "color": "#fff",
    "show_title": true
  }
}",
      "name": "yandex-browser-manifest.json",
    },
  ],
  "html": [
    "<link rel="icon" type="image/x-icon" href="https://domain/subdomain/favicon.ico">",
    "<link rel="icon" type="image/png" sizes="16x16" href="https://domain/subdomain/favicon-16x16.png">",
    "<link rel="icon" type="image/png" sizes="32x32" href="https://domain/subdomain/favicon-32x32.png">",
    "<link rel="icon" type="image/png" sizes="48x48" href="https://domain/subdomain/favicon-48x48.png">",
    "<link rel="manifest" href="https://domain/subdomain/manifest.webmanifest">",
    "<meta name="mobile-web-app-capable" content="yes">",
    "<meta name="theme-color" content="#fff">",
    "<meta name="application-name">",
    "<link rel="apple-touch-icon" sizes="57x57" href="https://domain/subdomain/apple-touch-icon-57x57.png">",
    "<link rel="apple-touch-icon" sizes="60x60" href="https://domain/subdomain/apple-touch-icon-60x60.png">",
    "<link rel="apple-touch-icon" sizes="72x72" href="https://domain/subdomain/apple-touch-icon-72x72.png">",
    "<link rel="apple-touch-icon" sizes="76x76" href="https://domain/subdomain/apple-touch-icon-76x76.png">",
    "<link rel="apple-touch-icon" sizes="114x114" href="https://domain/subdomain/apple-touch-icon-114x114.png">",
    "<link rel="apple-touch-icon" sizes="120x120" href="https://domain/subdomain/apple-touch-icon-120x120.png">",
    "<link rel="apple-touch-icon" sizes="144x144" href="https://domain/subdomain/apple-touch-icon-144x144.png">",
    "<link rel="apple-touch-icon" sizes="152x152" href="https://domain/subdomain/apple-touch-icon-152x152.png">",
    "<link rel="apple-touch-icon" sizes="167x167" href="https://domain/subdomain/apple-touch-icon-167x167.png">",
    "<link rel="apple-touch-icon" sizes="180x180" href="https://domain/subdomain/apple-touch-icon-180x180.png">",
    "<link rel="apple-touch-icon" sizes="1024x1024" href="https://domain/subdomain/apple-touch-icon-1024x1024.png">",
    "<meta name="apple-mobile-web-app-capable" content="yes">",
    "<meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">",
    "<meta name="apple-mobile-web-app-title">",
    "<link rel="apple-touch-startup-image" media="(device-width: 320px) and (device-height: 568px) and (-webkit-device-pixel-ratio: 2) and (orientation: portrait)" href="https://domain/subdomain/apple-touch-startup-image-640x1136.png">",
    "<link rel="apple-touch-startup-image" media="(device-width: 320px) and (device-height: 568px) and (-webkit-device-pixel-ratio: 2) and (orientation: landscape)" href="https://domain/subdomain/apple-touch-startup-image-1136x640.png">",
    "<link rel="apple-touch-startup-image" media="(device-width: 375px) and (device-height: 667px) and (-webkit-device-pixel-ratio: 2) and (orientation: portrait)" href="https://domain/subdomain/apple-touch-startup-image-750x1334.png">",
    "<link rel="apple-touch-startup-image" media="(device-width: 375px) and (device-height: 667px) and (-webkit-device-pixel-ratio: 2) and (orientation: landscape)" href="https://domain/subdomain/apple-touch-startup-image-1334x750.png">",
    "<link rel="apple-touch-startup-image" media="(device-width: 375px) and (device-height: 812px) and (-webkit-device-pixel-ratio: 3) and (orientation: portrait)" href="https://domain/subdomain/apple-touch-startup-image-1125x2436.png">",
    "<link rel="apple-touch-startup-image" media="(device-width: 375px) and (device-height: 812px) and (-webkit-device-pixel-ratio: 3) and (orientation: landscape)" href="https://domain/subdomain/apple-touch-startup-image-2436x1125.png">",
    "<link rel="apple-touch-startup-image" media="(device-width: 390px) and (device-height: 844px) and (-webkit-device-pixel-ratio: 3) and (orientation: portrait)" href="https://domain/subdomain/apple-touch-startup-image-1170x2532.png">",
    "<link rel="apple-touch-startup-image" media="(device-width: 390px) and (device-height: 844px) and (-webkit-device-pixel-ratio: 3) and (orientation: landscape)" href="https://domain/subdomain/apple-touch-startup-image-2532x1170.png">",
    "<link rel="apple-touch-startup-image" media="(device-width: 414px) and (device-height: 896px) and (-webkit-device-pixel-ratio: 2) and (orientation: portrait)" href="https://domain/subdomain/apple-touch-startup-image-828x1792.png">",
    "<link rel="apple-touch-startup-image" media="(device-width: 414px) and (device-height: 896px) and (-webkit-device-pixel-ratio: 2) and (orientation: landscape)" href="https://domain/subdomain/apple-touch-startup-image-1792x828.png">",
    "<link rel="apple-touch-startup-image" media="(device-width: 414px) and (device-height: 896px) and (-webkit-device-pixel-ratio: 3) and (orientation: portrait)" href="https://domain/subdomain/apple-touch-startup-image-1242x2688.png">",
    "<link rel="apple-touch-startup-image" media="(device-width: 414px) and (device-height: 896px) and (-webkit-device-pixel-ratio: 3) and (orientation: landscape)" href="https://domain/subdomain/apple-touch-startup-image-2688x1242.png">",
    "<link rel="apple-touch-startup-image" media="(device-width: 414px) and (device-height: 736px) and (-webkit-device-pixel-ratio: 3) and (orientation: portrait)" href="https://domain/subdomain/apple-touch-startup-image-1242x2208.png">",
    "<link rel="apple-touch-startup-image" media="(device-width: 414px) and (device-height: 736px) and (-webkit-device-pixel-ratio: 3) and (orientation: landscape)" href="https://domain/subdomain/apple-touch-startup-image-2208x1242.png">",
    "<link rel="apple-touch-startup-image" media="(device-width: 428px) and (device-height: 926px) and (-webkit-device-pixel-ratio: 3) and (orientation: portrait)" href="https://domain/subdomain/apple-touch-startup-image-1284x2778.png">",
    "<link rel="apple-touch-startup-image" media="(device-width: 428px) and (device-height: 926px) and (-webkit-device-pixel-ratio: 3) and (orientation: landscape)" href="https://domain/subdomain/apple-touch-startup-image-2778x1284.png">",
    "<link rel="apple-touch-startup-image" media="(device-width: 768px) and (device-height: 1024px) and (-webkit-device-pixel-ratio: 2) and (orientation: portrait)" href="https://domain/subdomain/apple-touch-startup-image-1536x2048.png">",
    "<link rel="apple-touch-startup-image" media="(device-width: 768px) and (device-height: 1024px) and (-webkit-device-pixel-ratio: 2) and (orientation: landscape)" href="https://domain/subdomain/apple-touch-startup-image-2048x1536.png">",
    "<link rel="apple-touch-startup-image" media="(device-width: 810px) and (device-height: 1080px) and (-webkit-device-pixel-ratio: 2) and (orientation: portrait)" href="https://domain/subdomain/apple-touch-startup-image-1620x2160.png">",
    "<link rel="apple-touch-startup-image" media="(device-width: 810px) and (device-height: 1080px) and (-webkit-device-pixel-ratio: 2) and (orientation: landscape)" href="https://domain/subdomain/apple-touch-startup-image-2160x1620.png">",
    "<link rel="apple-touch-startup-image" media="(device-width: 834px) and (device-height: 1194px) and (-webkit-device-pixel-ratio: 2) and (orientation: portrait)" href="https://domain/subdomain/apple-touch-startup-image-1668x2388.png">",
    "<link rel="apple-touch-startup-image" media="(device-width: 834px) and (device-height: 1194px) and (-webkit-device-pixel-ratio: 2) and (orientation: landscape)" href="https://domain/subdomain/apple-touch-startup-image-2388x1668.png">",
    "<link rel="apple-touch-startup-image" media="(device-width: 834px) and (device-height: 1112px) and (-webkit-device-pixel-ratio: 2) and (orientation: portrait)" href="https://domain/subdomain/apple-touch-startup-image-1668x2224.png">",
    "<link rel="apple-touch-startup-image" media="(device-width: 834px) and (device-height: 1112px) and (-webkit-device-pixel-ratio: 2) and (orientation: landscape)" href="https://domain/subdomain/apple-touch-startup-image-2224x1668.png">",
    "<link rel="apple-touch-startup-image" media="(device-width: 1024px) and (device-height: 1366px) and (-webkit-device-pixel-ratio: 2) and (orientation: portrait)" href="https://domain/subdomain/apple-touch-startup-image-2048x2732.png">",
    "<link rel="apple-touch-startup-image" media="(device-width: 1024px) and (device-height: 1366px) and (-webkit-device-pixel-ratio: 2) and (orientation: landscape)" href="https://domain/subdomain/apple-touch-startup-image-2732x2048.png">",
    "<meta name="msapplication-TileColor" content="#fff">",
    "<meta name="msapplication-TileImage" content="https://domain/subdomain/mstile-144x144.png">",
    "<meta name="msapplication-config" content="https://domain/subdomain/browserconfig.xml">",
    "<link rel="yandex-tableau-widget" href="https://domain/subdomain/yandex-browser-manifest.json">",
  ],
  "images": [
    {
      "contents": null,
      "name": "favicon.ico",
    },
    {
      "contents": null,
      "name": "favicon-16x16.png",
    },
    {
      "contents": null,
      "name": "favicon-32x32.png",
    },
    {
      "contents": null,
      "name": "favicon-48x48.png",
    },
    {
      "contents": null,
      "name": "android-chrome-36x36.png",
    },
    {
      "contents": null,
      "name": "android-chrome-48x48.png",
    },
    {
      "contents": null,
      "name": "android-chrome-72x72.png",
    },
    {
      "contents": null,
      "name": "android-chrome-96x96.png",
    },
    {
      "contents": null,
      "name": "android-chrome-144x144.png",
    },
    {
      "contents": null,
      "name": "android-chrome-192x192.png",
    },
    {
      "contents": null,
      "name": "android-chrome-256x256.png",
    },
    {
      "contents": null,
      "name": "android-chrome-384x384.png",
    },
    {
      "contents": null,
      "name": "android-chrome-512x512.png",
    },
    {
      "contents": null,
      "name": "apple-touch-icon-57x57.png",
    },
    {
      "contents": null,
      "name": "apple-touch-icon-60x60.png",
    },
    {
      "contents": null,
      "name": "apple-touch-icon-72x72.png",
    },
    {
      "contents": null,
      "name": "apple-touch-icon-76x76.png",
    },
    {
      "contents": null,
      "name": "apple-touch-icon-114x114.png",
    },
    {
      "contents": null,
      "name": "apple-touch-icon-120x120.png",
    },
    {
      "contents": null,
      "name": "apple-touch-icon-144x144.png",
    },
    {
      "contents": null,
      "name": "apple-touch-icon-152x152.png",
    },
    {
      "contents": null,
      "name": "apple-touch-icon-167x167.png",
    },
    {
      "contents": null,
      "name": "apple-touch-icon-180x180.png",
    },
    {
      "contents": null,
      "name": "apple-touch-icon-1024x1024.png",
    },
    {
      "contents": null,
      "name": "apple-touch-icon.png",
    },
    {
      "contents": null,
      "name": "apple-touch-icon-precomposed.png",
    },
    {
      "contents": null,
      "name": "apple-touch-startup-image-640x1136.png",
    },
    {
      "contents": null,
      "name": "apple-touch-startup-image-1136x640.png",
    },
    {
      "contents": null,
      "name": "apple-touch-startup-image-750x1334.png",
    },
    {
      "contents": null,
      "name": "apple-touch-startup-image-1334x750.png",
    },
    {
      "contents": null,
      "name": "apple-touch-startup-image-1125x2436.png",
    },
    {
      "contents": null,
      "name": "apple-touch-startup-image-2436x1125.png",
    },
    {
      "contents": null,
      "name": "apple-touch-startup-image-1170x2532.png",
    },
    {
      "contents": null,
      "name": "apple-touch-startup-image-2532x1170.png",
    },
    {
      "contents": null,
      "name": "apple-touch-startup-image-828x1792.png",
    },
    {
      "contents": null,
      "name": "apple-touch-startup-image-1792x828.png",
    },
    {
      "contents": null,
      "name": "apple-touch-startup-image-1242x2688.png",
    },
    {
      "contents": null,
      "name": "apple-touch-startup-image-2688x1242.png",
    },
    {
      "contents": null,
      "name": "apple-touch-startup-image-1242x2208.png",
    },
    {
      "contents": null,
      "name": "apple-touch-startup-image-2208x1242.png",
    },
    {
      "contents": null,
      "name": "apple-touch-startup-image-1284x2778.png",
    },
    {
      "contents": null,
      "name": "apple-touch-startup-image-2778x1284.png",
    },
    {
      "contents": null,
      "name": "apple-touch-startup-image-1536x2048.png",
    },
    {
      "contents": null,
      "name": "apple-touch-startup-image-2048x1536.png",
    },
    {
      "contents": null,
      "name": "apple-touch-startup-image-1620x2160.png",
    },
    {
      "contents": null,
      "name": "apple-touch-startup-image-2160x1620.png",
    },
    {
      "contents": null,
      "name": "apple-touch-startup-image-1668x2388.png",
    },
    {
      "contents": null,
      "name": "apple-touch-startup-image-2388x1668.png",
    },
    {
      "contents": null,
      "name": "apple-touch-startup-image-1668x2224.png",
    },
    {
      "contents": null,
      "name": "apple-touch-startup-image-2224x1668.png",
    },
    {
      "contents": null,
      "name": "apple-touch-startup-image-2048x2732.png",
    },
    {
      "contents": null,
      "name": "apple-touch-startup-image-2732x2048.png",
    },
    {
      "contents": null,
      "name": "mstile-70x70.png",
    },
    {
      "contents": null,
      "name": "mstile-144x144.png",
    },
    {
      "contents": null,
      "name": "mstile-150x150.png",
    },
    {
      "contents": null,
      "name": "mstile-310x150.png",
    },
    {
      "contents": null,
      "name": "mstile-310x310.png",
    },
    {
      "contents": null,
      "name": "yandex-browser-50x50.png",
    },
  ],
}
`;
